### PR TITLE
fix(cli/repl): interpret object literals as expressions

### LINKED ISSUE #1421: REPL emits surprising output on `> {a: 1}`
Currently,
```
> {a:1}
1
>
```
Expected:
```
> {a:1}
{ a: 1 }
>
```
This is due to v8 considering `{}` as a scope instead of object.
In Node, this is solved by secretly adding a pair of parentheses around the object, like
```
{a:1} => ({a:1})
```
The code implementing this is available here:
https://github.com/nodejs/node/blob/ae73b73eeb99101188c860cce488ccc085b2f268/lib/repl.js#L235-L243

I am expecting a similar patch for this after #1407 and #1420 are landed...

### PR DESCRIPTION
**Pull request by caspervonb**
This wraps object literals in parenthesis so that they're interpreted as expressions rather than block statements which matches the behaviour of Node and web-browsers.

Fixes #1421

### CODE REVIEW

**Comment by caspervonb** [comment]
Old behavior has some interesting edge cases

Master:

Screenshot 2020-09-20 at 1 59 22 PM

This pull request:

Screenshot 2020-09-20 at 2 01 41 PM

File: cli/rt/40_repl.js [hunk 1]
```diff
@@ -53,6 +53,13 @@
   // Returns true if code is consumed (no error/irrecoverable error).
   // Returns false if error is recoverable
   function evaluate(code) {
+    // It is a bit unexpected that { "foo": "bar" } is interpreted as a block
+    // statement rather than an object literal so we interpret it as an expression statement
+    // to match the behavior found in a typical prompt including browser developer tools.
+    if (code.startsWith("{") && code.endsWith("}")) {
+      code = `(${code.trim()})`;
+    }
```
Comment (nayeemrmn) [inline cli/rt/40_repl.js:61]:
```suggestion
    code = code.trim();
    if (code.startsWith("{") && code.endsWith("}")) {
      code = `(${code})`;
    }
```

**Review by nayeemrmn — COMMENTED** [review]
Fixes #1421. I suppose this is better even if block statements can't be used.

**Comment by kitsonk** [comment]
The existing behaviour is consistent with Node.js.  Is there compelling reason to vary from that?

**Comment by lucacasonato** [comment]
> The existing behaviour is consistent with Node.js. Is there compelling reason to vary from that?

It is not. 
image

**Comment by kitsonk** [comment]
Ah, but `{ "" }` does not throw a syntax error though.

**Comment by kitsonk** [comment]
Also, `{ "" }` in the dev console in Chrome doesn't throw either.  It returns `""`.

Maybe a RegEx that is a bit more specific in looking for things that look like an object literal but ignores things that couldn't be an object literal.  The following:

```js
/^\s*{\s*(?:\S+|["'][^"']+["']):.+}\s*$/
```

Matches:

```js
{ "foo": bar }
{ 'foo bar': bar }
{ [foo]: bar }
```

But ignores:

```js
{ "" }
```

It isn't perfect, but it is really hard to make it perfect because of the lack of back references in JavaScript.

**Comment by caspervonb** [comment]
> The existing behaviour is consistent with Node.js. Is there compelling reason to vary from that?

Node does the same, but with a different pattern match.

> Also, { "" } in the dev console in Chrome doesn't throw either. It returns "".

Throws in Safari, not sure about Firefox (not installed at the moment).

> /^\s*{\s*(?:\S+|["'][^"']+["']):.+}\s*$/

Yeah that regex is going to be pretty hard to get right, it should allow for non-quoted keys, but non-quoted keys look like labels, et cetera.

I think it might be better to have a rollback mechanism, e.g if our preprocessed code turned out to be a syntax error, do it again without the preprocessing.

**Comment by caspervonb** [comment]
Slightly more forgiving approach that leaves the parsing up to the compiler; if it looks like it might be an object literal we try it as an expression first, if thats a compile error we rollback and retry with the raw original code.

**Review by ry — APPROVED** [review]
Cool - nice hack.

**Pull request merged**